File: week2/Six.cpp
/* ************************************************************************
> File Name:     Six.cpp
> Author:        Zhenyu Song
> Mail:          zhenyus4@uci.edu
> Created Time:  Mon Oct 13 13:58:21 2025
> Description:   
 ************************************************************************/

#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <unordered_map>
#include <unordered_set>
#include <algorithm>
#include <cctype>
#include <sstream>

const std::string STOP_WORDS_PATH = "../stop_words.txt";
const int MAX_COUNT = 25;

std::unordered_set<std::string> loadStopWords(const std::string& filename) {
    std::unordered_set<std::string> stopWords;
    std::ifstream file(filename);
    if (!file.is_open()) {
        std::cerr << "failed to open " << filename << std::endl;
        return stopWords;
    }

    std::string line;
    while (std::getline(file, line)) {
        std::stringstream ss(line);
        std::string word;
        while (std::getline(ss, word, ',')) {
            stopWords.insert(word);
        }
    }
    file.close();
    return stopWords;
}

std::string readFile(const std::string& filename) {
    std::ifstream inputFile(filename);
    if (!inputFile.is_open()) {
        std::cerr << "failed to open " << filename << std::endl;
        return "";
    }
    std::stringstream buffer;
    buffer << inputFile.rdbuf();
    return buffer.str();
}

std::vector<std::string> filter(const std::string& data, std::unordered_set<std::string> stopWords) {
    std::vector<std::string> words;
    if (data.empty()) return words;

    auto dummyLine = data + " ";
    std::string currentWord;
    for (char ch : dummyLine) {
        if (isalnum(ch)) {
            currentWord += std::tolower(ch);
        } else if (!currentWord.empty()) {
            if (currentWord.length() >= 2 && !stopWords.count(currentWord)) {
                words.push_back(currentWord);
            }
            currentWord.clear();
        }
    }
    return words;
}

std::unordered_map<std::string, int> count(const std::vector<std::string>& words) {
    std::unordered_map<std::string, int> wordCounts;
    for (const auto& word : words) {
        ++wordCounts[word];
    }
    return wordCounts;
}

std::vector<std::pair<std::string, int>> sort(const std::unordered_map<std::string, int>& wordCounts) {
    std::vector<std::pair<std::string, int>> wordVec(wordCounts.begin(), wordCounts.end());
    std::sort(wordVec.begin(), wordVec.end(), [](const std::pair<std::string, int>& a, const std::pair<std::string, int>& b) {
        return a.second > b.second;
    });
    return wordVec;
}

void print(const std::vector<std::pair<std::string, int>>& sortedWords) {
    const int numToPrint = std::min(MAX_COUNT, static_cast<int>(sortedWords.size()));
    for (int i = 0; i < numToPrint; ++i) {
        std::cout << sortedWords[i].first << " - " << sortedWords[i].second << std::endl;
    }
}

int main(int argc, char* argv[]) {
    if (argc != 2) {
        std::cerr << "wrong usage, please enter: " << argv[0] << " <filename>" << std::endl;
        return 1;
    }

    print(
        sort(
            count(
                filter(
                    readFile(argv[1]),
                    loadStopWords(STOP_WORDS_PATH)
                )
            )
        )
    );

    return 0;
}

File: week2/Six.go
/* ************************************************************************
> File Name:     Six.go
> Author:        Zhenyu Song
> Mail:          zhenyus4@uci.edu
> Created Time:  Mon Oct 13 14:15:27 2025
> Description:   
 ************************************************************************/

package main

import (
	"fmt"
	"log"
	"os"
	"regexp"
	"sort"
	"strings"
)

const (
	stopWordsPath = "../stop_words.txt"
	maxCount      = 25
)

type Pair struct {
	Word  string
	Count int
}

func loadStopWords(filename string) map[string]bool {
	stopWords := make(map[string]bool)
	content, err := os.ReadFile(filename)
	if err != nil {
		log.Fatalf("failed to open %s: %v", filename, err)
	}

	words := strings.Split(string(content), ",")
	for _, word := range words {
		if len(word) > 0 {
			stopWords[word] = true
		}
	}
	return stopWords
}

func readFile(filename string) string {
	content, err := os.ReadFile(filename)
	if err != nil {
		log.Fatalf("failed to open %s: %v", filename, err)
	}
	return string(content)
}

func filter(data string, stopWords map[string]bool) []string {
	if data == "" {
		return []string{}
	}

	reg := regexp.MustCompile("[a-zA-Z0-9]+")
	words := reg.FindAllString(data, -1)

	filteredWords := []string{}
	for _, word := range words {
		lowerWord := strings.ToLower(word)
		if len(lowerWord) >= 2 {
			if _, isStopWord := stopWords[lowerWord]; !isStopWord {
				filteredWords = append(filteredWords, lowerWord)
			}
		}
	}
	return filteredWords
}

func count(words []string) map[string]int {
	wordCounts := make(map[string]int)
	for _, word := range words {
		wordCounts[word]++
	}
	return wordCounts
}

func sortWords(wordCounts map[string]int) []Pair {
	pairs := make([]Pair, 0, len(wordCounts))
	for word, count := range wordCounts {
		pairs = append(pairs, Pair{word, count})
	}

	sort.Slice(pairs, func(i, j int) bool {
		return pairs[i].Count > pairs[j].Count
	})

	return pairs
}

func printWords(sortedWords []Pair) {
	numToPrint := maxCount
	if len(sortedWords) < maxCount {
		numToPrint = len(sortedWords)
	}

	for i := 0; i < numToPrint; i++ {
		fmt.Printf("%s - %d\n", sortedWords[i].Word, sortedWords[i].Count)
	}
}

func main() {
	if len(os.Args) != 2 {
		fmt.Fprintf(os.Stderr, "wrong usage, please enter: %s <filename>\n", os.Args[0])
		os.Exit(1)
	}

	printWords(
		sortWords(
			count(
				filter(
					readFile(os.Args[1]),
					loadStopWords(stopWordsPath),
				),
			),
		),
	)
}
